### DISCUSSION TITLE
Is the internal postgres schema defined?

### DISCUSSION THREAD

**Discussion by maxott** [geldata/gel#3598] [Q&A] [answered]
Hi,

I have been waiting for something like EdgeDB for quite some time. It's a very pragmatic approach to graph dbs, but also takes advantage of a battle-tested platform. And it is primarily the latter which would make it a lot easier for me to sell this to my team IF we could easily fall back to a Postgres only mode if something doesn't work out - for whatever reason.

So my question: Is the way EdgeDB stores data in postgres 'open' in the sense that there is some attempt to keep that stable at some stage, so that technically one could retrieve data from postgres directly if the EdgeDB has to be removed. 

Thanks,
-max

**Comment by msullivan** [comment, marked as answer]
Accessing the underlying postgres database is definitely possible, though I wouldn't exactly call it "supported".

We don't make any *promises* about not changing the layout of the underlying postgres database in future versions of edgedb, but also we are mostly pretty happy with how we store data in postgres and I don't know of any major changes planned. (One possible change I can think of is that we currently always store link properties in a link table, even when the link is single; we might change that to move the link property into the source table in that case.)

One thing to note is that everything is named based on the uuid `id`s of the corresponding schema objects instead of their actual names. You can find these `id`s by querying the schema introspection types in the `schema` database.

I would also caution against trying to *write* into a edgedb database manually from SQL; we don't declare any foreign keys, since our compiler ensures that all links are valid without needing to include the constraint on the postgres side. (If you were moving off edgedb and trying to keep using the same postgres database, you'd probably want to rename everything and add foreign keys.)

**Reply by maxott** [reply]
Thanks for the quick reply. Good to know and I fully understand that writing into postgres in parallel to edgedb is not a good idea. For me this discussion is primarily in regards to risk mitigation. 

Having said that, is there anywhere in the docs/code where I could learn a bit more on how the mapping to sql tables is done?

**Reply by msullivan** [reply]
Probably there should be, but I'll just sketch out the key parts here:
 * Every object type has a table, named after the `id` of the object type
 * Each `single` link/property has a column in the object table, named after the `id` of the link/property. (Except for `id` and `__type__`, which are just named that.)
 * Link columns contain the `id` of the target
 * Each `multi` link/property has a link table, named after the `id` of the link/property
   * That table has `source` and `target` columns
   * And also a column for each link property on the link, named after the `id` of the link prop
 * Link properties are always stored in a link table, so a `single link` that has link properties will have both a column in the object table *and* a link table. (This is the thing most likely to change--we may move such link properties into the object itself.)
 * For every object type table and link table named `{uuid}`, there is also a view named `{uuid}_t` which unions together that table and the tables for all of its descendants. Usually this is what you want when querying.

You can get all of these `id`s by querying `schema`. The following query I think gets all the information you need:
```
select schema::ObjectType {
    id,
    name,
    pointers := (select .pointers {
        [is schema::Link].properties: {id, name}
            filter .name not in {"source", "target"} and not exists.expr
    } {
        id,
        name,
        cardinality,
        is_inline := .cardinality = schema::Cardinality.One,
        has_table := .cardinality = schema::Cardinality.Many
                     or count(.properties) > 2,
        properties: {id, name}
   } filter .name not in {"id", "__type__"} and not exists .expr)
}
filter not .builtin and not .from_alias;
```
